semgrep-core/src/engine/Eval_generic.ml

```diff
@@ -304,33 +336,45 @@ let bindings_to_env (config : Config_semgrep.t) xs =
   in
   { mvars; constant_propagation }
 
-let text_of_binding mvar mval =
-  match mval with
-  | MV.Text (text, _) ->
-      (* Note that `text` may be produced by constant folding, in which
-       * case we will not have range info. *)
-      Some text
-  | _ -> (
-      let any = MV.mvalue_to_any mval in
-      match Visitor_AST.range_of_any_opt any with
-      | None ->
-          (* TODO: Report a warning to the user? *)
-          logger#error "We lack range info for metavariable %s: %s" mvar
-            (G.show_any any);
-          None
-      | Some (min, max) ->
-          let file = min.Parse_info.file in
-          let range = Range.range_of_token_locations min max in
-          Some (Range.content_at_range file range))
+let string_of_binding mvar mval =
+  Option.bind (text_of_binding mvar mval) @@ fun x -> Some (mvar, String x)
 
 (* this is for metavariable-regexp *)
-let bindings_to_env_with_just_strings (config : Config_semgrep.t) bindings =
-  let ( let* ) = Option.bind in
+let bindings_to_env_just_strings_const_prop (config : Config_semgrep.t) xs =
+  let constant_propagation = config.constant_propagation in
   let mvars =
-    bindings
+    xs
     |> Common.map_filter (fun (mvar, mval) ->
-           let* text = text_of_binding mvar mval in
-           Some (mvar, String text))
+           let try_bind_to_exp e =
```

Comment:
This looks like copy-pasted from `bindings_to_env`, can we refactor?

Reply:
yeah, it is. Definitely ugly. Initially I tried to combine them but the resulting function felt pretty hard to read. I'll give it another go before merging

Reply:
No worries, sometimes it's better to have duplicate code. I just wanted to point out that possibility of refactoring, but you did consider it already. 👍